File: activity/20200805/RabbitLee/Increasing_Order_Search_Tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def increasingBST(self, root: TreeNode) -> TreeNode:
        ans = self.cur = TreeNode()
        self.relink(root)
        return ans.right
        
    def relink(self, node: TreeNode) -> None:
        if node:
            self.relink(node.left)
            self.cur.left = None
            self.cur.right = node
            self.cur = node
            self.relink(node.right)

File: activity/20200805/scau-flyingbird/Increasing_Order_Search_Tree.go
/**
 * Definition for a binary tree node.
 */
type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

func walk(root *TreeNode, ch chan int) {
	if root == nil {
		return
	}

	walk(root.Left, ch)
	ch <- root.Val
	walk(root.Right, ch)
}

func increasingBST(root *TreeNode) *TreeNode {
	ch := make(chan int)
	go func() {
		defer close(ch)
		walk(root, ch)
	}()

	ordered := &TreeNode{}
	head := ordered
	for c := range ch {
		t := &TreeNode{c, nil, nil}
		ordered.Right = t
		ordered = ordered.Right
	}
	return head
}
